privacy/bolton/model.py

```diff
@@ -0,0 +1,551 @@
+"""Bolton model for bolton method of differentially private ML"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+import tensorflow as tf
+from tensorflow.python.keras.models import Model
+from tensorflow.python.keras import optimizers
+from tensorflow.python.framework import ops as _ops
+from privacy.bolton.loss import StrongConvexMixin
+from privacy.bolton.optimizer import Bolton
+
+_accepted_distributions = ['laplace']
+
+
+class BoltonModel(Model):
+  """
+  Bolton episilon-delta model
+  Uses 4 key steps to achieve privacy guarantees:
+  1. Adds noise to weights after training (output perturbation).
+  2. Projects weights to R after each batch
+  3. Limits learning rate
+  4. Use a strongly convex loss function (see compile)
+
+  For more details on the strong convexity requirements, see:
+  Bolt-on Differential Privacy for Scalable Stochastic Gradient
+  Descent-based Analytics by Xi Wu et. al.
+  """
+
+  def __init__(self,
+               n_classes,
+               # noise_distribution='laplace',
+               seed=1,
+               dtype=tf.float32
+               ):
+    """ private constructor.
+
+    Args:
+        n_classes: number of output classes to predict.
+        epsilon: level of privacy guarantee
+        noise_distribution: distribution to pull weight perturbations from
+        weights_initializer: initializer for weights
+        seed: random seed to use
+        dtype: data type to use for tensors
+    """
+
+    # if noise_distribution not in _accepted_distributions:
+    #   raise ValueError('Detected noise distribution: {0} not one of: {1} valid'
+    #                    'distributions'.format(noise_distribution,
+    #                                           _accepted_distributions))
+    # if epsilon <= 0:
+    #   raise ValueError('Detected epsilon: {0}. '
+    #                    'Valid range is 0 < epsilon <inf'.format(epsilon))
+    # self.epsilon = epsilon
+    super(BoltonModel, self).__init__(name='bolton', dynamic=False)
+    self.n_classes = n_classes
+    self.force = False
+    # self.noise_distribution = noise_distribution
+    self.seed = seed
+    self.__in_fit = False
+    self._layers_instantiated = False
+    # self._callback = MyCustomCallback()
+    self._dtype = dtype
+
+  def call(self, inputs):
+    """Forward pass of network
+
+    Args:
+        inputs: inputs to neural network
+
+    Returns:
+
+    """
+    return self.output_layer(inputs)
+
+  def compile(self,
+              optimizer='SGD',
+              loss=None,
+              metrics=None,
+              loss_weights=None,
+              sample_weight_mode=None,
+              weighted_metrics=None,
+              target_tensors=None,
+              distribute=None,
+              **kwargs):
+    """See super class. Default optimizer used in Bolton method is SGD.
+
+    """
+    for key, val in StrongConvexMixin.__dict__.items():
```

Comment:
We should be able to enforce this using a more natural approach. We can declare the methods in StrongConvexMixin to be abstract method and then check the loss passed in is derived from StronglyConvexMixin.